internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,131 @@
+	// ContainerCredentialGuardAddInstanceRequest. This is where the cred spec
+	// gets passed in. Transport either "LRPC" (Argon) or "HvSocket" (Xenon).
+	// 2. Query the instance with a call to HcsGetServiceProperties with the
+	// PropertyType "ContainerCredentialGuard". This will return all instances
+	// 3. Parse for the id of our container to find which one correlates to the
+	// container we're building the doc for, then add to the v2 doc.
+	// 4. If xenon container the hvsocketconfig will need to be attached BEFORE
+	// being created. It must be in the doc itself as we do not support hot adding
+	// service table entries. This is currently a blocker for adding support for
+	// hyper-v gmsa.
+	transport := "LRPC"
+	if uvm {
+		transport = "HvSocket"
+	}
+	req := hcsschema.ModificationRequest{
+		PropertyType: hcsschema.PTContainerCredentialGuard,
+		Settings: &hcsschema.ContainerCredentialGuardOperationRequest{
+			Operation: hcsschema.AddInstance,
+			OperationDetails: &hcsschema.ContainerCredentialGuardAddInstanceRequest{
+				Id:             ID,
+				CredentialSpec: credSpec,
+				Transport:      transport,
+			},
+		},
+	}
+	if err := hcs.ModifyServiceSettings(ctx, req); err != nil {
+		log.G(ctx).WithError(err).WithField("containerID", ID).Warn("failed to generate container credential guard instance")
+		return nil, err
+	}
+
+	q := hcsschema.PropertyQuery{
+		PropertyTypes: []hcsschema.PropertyType{hcsschema.PTContainerCredentialGuard},
+	}
+	serviceProps, err := hcs.GetServiceProperties(ctx, q)
+	if err != nil {
+		log.G(ctx).WithError(err).WithField("containerID", ID).Warn("failed to retrieve container credential guard instances")
+		return nil, err
+	}
+
+	ccgState, err := unmarshalCCGInstances(serviceProps, ID)
+	if err != nil {
+		return nil, err
+	}
+	return ccgState, nil
+}
+
+// Checks to see if doc has non-nil ccg field
+func hasCCG(doc *hcsschema.Container) bool {
+	if doc.ContainerCredentialGuard != nil {
+		return true
+	}
+	return false
+}
+
+// Removes a ContainerCredentialGuard instance by container ID.
+func removeCredentialGuard(ctx context.Context, ID string) error {
+	log.G(ctx).WithField("containerID", ID).Debug("removing container credential guard")
+
+	req := hcsschema.ModificationRequest{
+		PropertyType: hcsschema.PTContainerCredentialGuard,
+		Settings: &hcsschema.ContainerCredentialGuardOperationRequest{
+			Operation: hcsschema.RemoveInstance,
+			OperationDetails: &hcsschema.ContainerCredentialGuardRemoveInstanceRequest{
+				Id: ID,
+			},
+		},
+	}
+	if err := hcs.ModifyServiceSettings(ctx, req); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Unmarshals a ServiceProperties struct if the only property queried was 'ContainerCredentialGuard'
+// Returns the ContainerCredentialGuardInstance matching the id passed in.
+func unmarshalCCGInstances(sp *hcsschema.ServiceProperties, ID string) (*hcsschema.ContainerCredentialGuardState, error) {
```

Comment:
I don't think this is a good name for the function since it also does a search for the given id, it's not just unmarshalling. How about naming it `getCCGInstance`, and having it do the query call as well?

Reply:
I think I'll just leave this to do the unmarshalling and do the ID search in CreateCredentialGuard after this call as its pretty straightforward

Reply:
Should we just remove `unmarshalCCGInstances` then? It's not called anywhere else.

Reply:
After removing the id search yea its even more useless. I'll just do it all in CreateCredentialGuard